File: KthSmallestElementBST_recursive.py
# TC: O(K) since we will be not be traversing elements greater than K elements. 
# SC: O(H) where H is the size of the recursive stack.

class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        self.count = 0
        self.result = -1
        def inorder(root, k): 
#             base
            if not root: 
                return
        
#             logic
            inorder(root.left, k)
            self.count += 1
            if self.count == k:
                self.result = root.val
            
            if self.result == -1: 
                inorder(root.right, k)
            
        inorder(root, k)
        return self.result

File: kthSmallestElementBST_Recursive.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    
    int x;
    
    public int KthSmallest(TreeNode root, int k) {
        
        x = k;
        return Helper(root, k);
    }
    
    private int Helper(TreeNode root, int k) {
        
        if(root == null) {
            return Int32.MaxValue;
        }
        
        int l = Helper(root.left, k);
       
        x--;
        
        if(x == 0) {
            return root.val;
        }
        
        int r = Helper(root.right, x);
        
        return Math.Min(l, r);
    }
    
}